### ISSUE TITLE
Rooms: Live refactor

### ISSUE THREAD

**Issue by drunkplato** [holium/realm#638]
A presence lib and agents for Realm and Realm-enabled applications.

### `room`
```hoon
::
+$  room
  $:  rid=@t
      provider=ship
      creator=ship
      access=?(%public %private)
      title=cord
      present=(set ship)
      whitelist=(set ship)
      capacity=@ud
      path=(unit cord)
      metadata=(cord cord)
  ==
::
+$  rooms  (map rid room)
```

### Top level state
- `live`: the state related to your present room.
- `context`: the state related to the context of your rooms list
- `provider`: the state related to the rooms list you actually provide for others.

#### `live`
```hoon
+$  live  
  $:  =room
      stack=(unit stack-id)
  ==
```

#### `context`
```hoon
+$  context  
  $:  =rooms
      provider=ship 
  ==
```

#### `provider`
```hoon
+$  provider  
  $:  =rooms
      online=?
      banned=(set ship)
  ==
```

#### `state-1`
```hoon
+$  state-1
  $:  %1
      =live
      =context
      =provider
  ==
```

## The agents

- `%rooms`: signaling and room management agent 
- `%room-session`: the multiplayer desktop backend (future work)

## `@holium/realm-rooms`

This lib allows for developers to interface with the rooms primitive.

There are a few concepts to detail here.

TODO

## Latency handling pre-signaling

```
HIGHER      |  LOWER
------------|----------
            |<- ready
ack-ready ->| 
```

Once the pre-signaling handshake is complete,  WebRTC auto-connection begins. This is handled by the [`simple-peer`](https://github.com/feross/simple-peer) lib.

## Auto-connection webrtc protocol

```
HIGHER  |  LOWER
--------|----------
        |<- awaiting offer
offer ->| 
        |<- answer
        |   
  ice <---> ice 
        |   
```

first, the lower @p sends `awaiting-offer` (as many times as it takes)

when the higher @p receives `awaiting-offer`, it returns an `answer`

when the lower @p receives the `offer` it stops sending `awaiting-offer`, processes the `offer` and returns an `answer`

when either party has processes the others `offer|answer`, it can begin processing ice candidates.
